File: src/p0004/python/solution.py
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        total = len(nums1) + len(nums2)
        kth = (total + 1) // 2
        begin = max(kth, len(nums2)) - len(nums2)
        end = min(kth, len(nums1))
        while begin < end:
            i = begin + (end - begin) // 2
            j = kth - i
            if nums1[i] - nums2[j - 1] < 0:
                begin = i + 1
            else:
                end = i
        i = begin
        j = kth - i
        if j == 0:
            left = nums1[i - 1]
        elif i == 0:
            left = nums2[j - 1]
        else:
            left = max(nums1[i - 1], nums2[j - 1])
        if total % 2 != 0:
            return left
        if j == len(nums2):
            right = nums1[i]
        elif i == len(nums1):
            right = nums2[j]
        else:
            right = min(nums1[i], nums2[j])
        return (left + right) / 2

File: src/p0004/java/solution.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int total = nums1.length + nums2.length;
        int kth = (total + 1) / 2;
        int begin = Math.max(kth, nums2.length) - nums2.length;
        int end = Math.min(kth, nums1.length);
        while (begin < end) {
            int i = begin + (end - begin) / 2;
            int j = kth - i;
            if (nums1[i] - nums2[j - 1] < 0) {
                begin = i + 1;
            } else {
                end = i;
            }
        }
        int i = begin;
        int j = kth - i;
        double left;
        if (j == 0) {
            left = nums1[i - 1];
        } else if (i == 0) {
            left = nums2[j - 1];
        } else {
            left = Math.max(nums1[i - 1], nums2[j - 1]);
        }
        if (total % 2 != 0) {
            return left;
        }
        double right;
        if (j == nums2.length) {
            right = nums1[i];
        } else if (i == nums1.length) {
            right = nums2[j];
        } else {
            right = Math.min(nums1[i], nums2[j]);
        }
        return (left + right) / 2;
    }
}
